### PR TITLE
tests: Refactor TestBundle and LoadTracking

### PR DESCRIPTION
**Pull request by douglas-raillard-arm**
* Refactor TestBundle and RTATestBundle to split the test aspects from
the from_testenv aspect. This is needed for some tests that create a
collection of bundles.

* Refactor the *Invariance test in load_tracking to enable test_slack
and avoid code duplication

### CODE REVIEW

File: doc/internals.rst [hunk 1]
```diff
@@ -18,6 +18,17 @@ be contributed to Devlib instead.
 .. automodule:: lisa.env
    :members:
 
+PlatformInfo
```
Comment (valschneider) [inline doc/internals.rst]: This change belongs in a separate commit.

File: lisa/exekall_customize.py [hunk 2]
```diff
@@ -177,15 +178,21 @@ def _finalize_expr_val(self, expr_val, artifact_dir, testcase_artifact_dir):
 
     @classmethod
     def get_tag_list(cls, value):
+        tags = []
```
Comment (valschneider) [inline lisa/exekall_customize.py]: Ditto about splitting changes.

File: lisa/tests/kernel/scheduler/load_tracking.py [hunk 3]
```diff
@@ -25,7 +26,7 @@
 
 from trappy.stats.Topology import Topology
 
-from lisa.tests.kernel.test_bundle import TestMetric, Result, ResultBundle, RTATestBundle
+from lisa.tests.kernel.test_bundle import TestMetric, Result, ResultBundle, RTATestBundleFromTestEnv, RTATestBundleTests, RTATestBundleBase, RTATestBundle
```
Comment (valschneider) [inline lisa/tests/kernel/scheduler/load_tracking.py]: That line is begging for a break

File: lisa/tests/kernel/test_bundle.py [hunk 4]
```diff
@@ -155,14 +154,61 @@ class CannotCreateError(RuntimeError):
     """
     pass
 
-class TestBundle(Serializable, abc.ABC):
+class TestBundleBase(Serializable):
```
Comment (valschneider) [inline lisa/tests/kernel/test_bundle.py]: Why create this split? I only had a quick look but I didn't find a user for this (there's LoadTrackingBase but it implements _from_testenv for some reason).

I also strongly believe all of our TestBundles should have the `from_testenv` method. It's an alternative constructor after all, if for some reason some ugly duckling doesn't want it (and I don't see a reason why we'd even have one like this), we can document/raise somewhere in there.

**Comment by valschneider** [comment]
Out of band discussion with @douglas-raillard-arm: We can get rid of the `TestBundle` split and have `FrequencyInvarianceItem`s be full-fledged `RTATestBundle`. The class bundling them up together (one per freq) can be a regular `TestBundle`.

**Comment by valschneider** [comment]
Yet another victim of the "close and comment" button that looks like the "comment" one...

File: lisa/exekall_customize.py (hunk 2 above)
Comment (douglas-raillard-arm) [reply lisa/exekall_customize.py]: this one does belong to here, I am adding tags to a type I introduce here

**Comment by valschneider** [comment]
I'm reviewing the code right now, but something came to mind - could you update the devlib subtree as well? That way we get the fixes to governor switching.

File: lisa/tests/kernel/scheduler/load_tracking.py [hunk 5]
```diff
@@ -337,73 +366,153 @@ def get_rtapp_profile(cls, te):
...
-        return cls(res_dir, te.plat_info, rtapp_profile, freqs)
+    @classmethod
+    def from_testenv(cls, te:TestEnv, res_dir:ArtifactPath=None) -> 'FreqInvariance':
+        return super().from_testenv(te, res_dir)
+
+    @classmethod
+    def _from_testenv(cls, te, res_dir):
+        return cls(res_dir, te.plat_info, OrderedDict(
+            (item.freq, item)
+            for item in self.iter_freq_items(te, res_dir)
```
Comment (valschneider) [inline lisa/tests/kernel/scheduler/load_tracking.py]:
```suggestion
            for item in cls.iter_freq_items(te, res_dir)
```

File: lisa/tests/kernel/scheduler/load_tracking.py [hunk 6]
```diff
@@ -337,73 +366,153 @@ def get_rtapp_profile(cls, te):
...
+    Hz to an instance of :class:`FreqInvarianceItem`.
+    """
+
+    def __init__(self, res_dir, plat_info, frequencies):
+        super().__init__(res_dir, plat_info)
+
+        self.frequencies = frequencies
+
+    @classmethod
+    def iter_freq_items(cls, te:TestEnv, res_dir:ArtifactPath) -> FreqInvarianceItem:
```
Comment (valschneider) [inline lisa/tests/kernel/scheduler/load_tracking.py]: A docstring here would be nice

File: lisa/tests/kernel/scheduler/load_tracking.py [hunk 7]
```diff
@@ -337,73 +366,153 @@ def get_rtapp_profile(cls, te):
...
+    cross section of available frequencies.
+
+    This class is mostly a wrapper around :class:`FreqInvarianceItem`,
+    providing a way to build a list of those for a few frequencies, and
+    providing aggregated versions of the tests. Calling the tests methods on
+    the items directly is recommended to avoid the unavoidable loss of
+    information when aggregating the
+    :class:`~lisa.tests.kernel.test_bundle.Result` of each item.
+
+    `frequencies` instance attribute is a mapping of frequency values in
```
[the remark this answers is not in the document]
Comment (douglas-raillard-arm) [reply lisa/tests/kernel/scheduler/load_tracking.py]: renamed to freq_items
